test/unit/Cardano/Wallet/CoinSelection/RandomSpec.hs

```diff
@@ -0,0 +1,290 @@
+                    $ NE.map Coin txOutsCoins
+            pure (utxo, txOuts)
+
+
+propFragmentation
+    :: CoveringCase
+    -> Property
+propFragmentation (CoveringCase (utxo, txOuts)) = do
+    isRight selection1 && isRight selection2 ==>
+        let (Right s1, Right s2) = trace ("selection1" <> show selection1 <> " selection2:"<> show selection2) $
+                (selection1, selection2)
+        in prop (s1, s2)
+  where
+    prop (CoinSelection inps1 _ _, CoinSelection inps2 _ _) =
+        L.length inps1 `shouldSatisfy` (>= L.length inps2)
+    selection1 = runIdentity $ runExceptT $
+        random (defaultCoinSelectionOptions 100) utxo txOuts
+    selection2 = runIdentity $ runExceptT $
+        largestFirst (defaultCoinSelectionOptions 100) utxo txOuts
+
+propErrors
+    :: CoveringCase
+    -> Property
+propErrors (CoveringCase (utxo, txOuts)) = do
+    isLeft selection1 && isLeft selection2 ==>
+        let (Left s1, Left s2) =
+                (selection1, selection2)
+        in prop (s1, s2)
+  where
+    prop (err1, err2) =
+        err1 === err2
+    selection1 = runIdentity $ runExceptT $
+        random (defaultCoinSelectionOptions 100) utxo txOuts
+    selection2 = runIdentity $ runExceptT $
+        largestFirst (defaultCoinSelectionOptions 100) utxo txOuts
+
+
+{-------------------------------------------------------------------------------
+                                  Test Data
+-------------------------------------------------------------------------------}
+
+newtype CoveringCase = CoveringCase { getCoveringCase :: (UTxO, NonEmpty TxOut)}
+    deriving Show
+
+instance Arbitrary CoveringCase where
+    arbitrary = do
+        n <- choose (1, 10)
+        txOutsNonEmpty <- NE.fromList <$> vectorOf n arbitrary
+        utxo <- arbitrary
+        return $ CoveringCase (utxo, txOutsNonEmpty)
+
+instance Arbitrary Address where
+    -- No Shrinking
+    arbitrary = oneof
+        [ pure $ Address "ADDR01"
+        , pure $ Address "ADDR02"
+        , pure $ Address "ADDR03"
+        ]
+
+instance Arbitrary Coin where
+    -- No Shrinking
+    arbitrary = Coin <$> choose (1, 100000)
+
+instance Arbitrary TxIn where
+    -- No Shrinking
+    arbitrary = TxIn
+        <$> arbitrary
+        <*> scale (`mod` 3) arbitrary -- No need for a high indexes
+
+instance Arbitrary (Hash "Tx") where
+    -- No Shrinking
+    arbitrary = do
+        wds <- vectorOf 10 arbitrary :: Gen [Word8]
+        let bs = BS.pack wds
+        pure $ Hash bs
+
+instance Arbitrary TxOut where
+    -- No Shrinking
+    arbitrary = TxOut
+        <$> arbitrary
+        <*> arbitrary
+
+instance Arbitrary UTxO where
+    shrink (UTxO utxo) = UTxO <$> shrink utxo
+    arbitrary = do
+        n <- choose (1, 100)
+        utxo <- zip
+            <$> vectorOf n arbitrary
+            <*> vectorOf n arbitrary
+        return $ UTxO $ Map.fromList utxo
+
+instance MonadRandom Identity where
```

Comment:
this is probably wrong

Reply:
it makes sense to me that this will loop forever. On the right side you are calling the same `getRandomBytes` defined in the `instance MonadRandom Identity` , with the same params. 

Looking at https://www.stackage.org/haddock/lts-13.15/cryptonite-0.25/Crypto-Random-Types.html what I believe you should have aimed for is to use one of two implementations of `class DRG` (from the same module) for implementation of `MonadRandom Identity` as you would like to use some pseudo random generator in this pure code.

I can have a closer look later and experiment if this doesn't give any clue (hope its the right one)

Reply:
thanks @akegalj for the comment. Indeed, random monad within random algo needs more effects power than the monad within largestFirst. there is a need of more effects to enable randomness - be in IO or have initial seed/generator.